src/adapters/Celer/CelerReceiverAdapter.sol

```diff
@@ -1,87 +1,139 @@
+    /// @dev tracks the msg id status to prevent replay
+    mapping(bytes32 => bool) public isMessageExecuted;
+
+    /*/////////////////////////////////////////////////////////////////
+                                 MODIFIER
+    ////////////////////////////////////////////////////////////////*/
+    modifier onlyCaller() {
+        if (!gac.isPrivilegedCaller(msg.sender)) {
+            revert Error.INVALID_PRIVILEGED_CALLER();
+        }
+        _;
+    }
 
     modifier onlyMessageBus() {
-        require(msg.sender == msgBus, "caller is not message bus");
+        if (msg.sender == msgBus) {
+            revert Error.CALLER_NOT_CELER_BUS();
+        }
         _;
     }
 
-    constructor(address _msgBus) {
+    /*/////////////////////////////////////////////////////////////////
+                                CONSTRUCTOR
+    ////////////////////////////////////////////////////////////////*/
+    constructor(address _msgBus, address _gac) {
         msgBus = _msgBus;
+        gac = IGAC(_gac);
+    }
+
+    /*/////////////////////////////////////////////////////////////////
+                                EXTERNAL FUNCTIONS
+    ////////////////////////////////////////////////////////////////*/
+
+    /// @inheritdoc IBridgeReceiverAdapter
+    function updateSenderAdapter(bytes memory _senderChain, address _senderAdapter) external override onlyCaller {
+        uint64 _senderChainDecoded = abi.decode(_senderChain, (uint64));
+
+        if (_senderChainDecoded == 0) {
+            revert Error.ZERO_CHAIN_ID();
+        }
+
+        if (_senderAdapter == address(0)) {
+            revert Error.ZERO_ADDRESS_INPUT();
+        }
+
+        address oldAdapter = senderAdapter;
+        senderAdapter = _senderAdapter;
+        senderChain = _senderChainDecoded;
+
+        emit SenderAdapterUpdated(oldAdapter, _senderAdapter, _senderChain);
     }
 
-    // Called by MessageBus on destination chain to receive cross-chain messages.
-    // The message is abi.encode of (MessageStruct.Message).
+    /// @dev accepts incoming messages from celer message bus
     function executeMessage(
         address _srcContract,
         uint64 _srcChainId,
         bytes calldata _message,
         address // executor
-    ) external payable override onlyMessageBus whenNotMsgPaused returns (ExecutionStatus) {
-        (bytes32 msgId, address srcSender, address destReceiver, bytes memory data) =
-            abi.decode(_message, (bytes32, address, address, bytes));
-        require(_srcContract == senderAdapters[uint256(_srcChainId)], "not allowed message sender");
-        if (executedMessages[msgId]) {
+    ) external payable override onlyMessageBus returns (ExecutionStatus) {
+        /// @dev step-1: validate incoming chain id
+        if (_srcChainId != senderChain) {
+            revert Error.INVALID_SENDER_CHAIN_ID();
+        }
+
+        /// @dev step-2: validate the caller (done in modifier)
```

Comment:
nit: this comment is misleading. suggest removing or moving it before comment on 101, as step 1